django/db/models/query.py

```diff
@@ -2451,7 +2451,14 @@ def has_to_attr_attribute(instance):
             # get_prefetch_queryset() method.
             if hasattr(rel_obj_descriptor, "get_prefetch_queryset"):
                 prefetcher = rel_obj_descriptor
-                is_fetched = rel_obj_descriptor.is_cached
+
+                # If to_attr has been set, then we simply check if the value
+                # has already been set, which is done by has_to_attr_attribute.
+                # We do not use the method from the descriptor, as the cache_name
+                # it defines checks with the field name, not with the to_attr
+                # value.
+                if through_attr == to_attr:
+                    is_fetched = rel_obj_descriptor.is_cached
```

Comment:
This makes sense to me particularly given how we special case `through_attr != to_attr:` in the many related manager case but not here.

I does make me wonder if we need similar logic to the one below wrt/ to cached properties though.

Maybe we could make `has_to_attr_attribute` more generic to handle both cases?

```python
def is_to_attr_fetched(model, to_attr):
    # Special case cached_property instances because hasattr
    # triggers attribute computation and assignment.
    if isinstance(
        getattr(model, to_attr, None), cached_property
    ):
        def has_cached_property(instance):
            return to_attr in instance.__dict__

        return has_cached_property

    def has_to_attr_attribute(instance):
        return hasattr(instance, to_attr)

    return has_to_attr_attribute

is_fetched = is_to_attr_fetched(instance.__class__, to_attr)
```

Reply:
I quite like this idea, I'll look into it and make the change.